### PR TITLE
fix: use configured RayCluster service account when autoscaling

### LINKED ISSUE #242: [Bug] The new SA used by autoscaler caused permission errors
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

ray-operator

### What happened + What you expected to happen

When I set `enableInTreeAutoscaling` to `true` for my Ray cluster, the kuberay-operator creates a new k8s service account with the same name of the cluster (if it's not available) and uses it for the head node pod (see [here](https://github.com/ray-project/kuberay/blob/master/ray-operator/controllers/raycluster_controller.go#L585)). Our Ray clusters on Kubernetes use a specific pre-configured service account in the namespace. Switching to a different service account would cause the permission issues when users try to use Ray to read the data. The autoscaler should use the existing service account defined in the `RayCluster` resource object for the head node instead of creating a new one to use.   

### Reproduction script

The bug can be reproduced by following these steps.
- Define a k8s service account named `default-sa`
- Create a RayCluster with the name of `foo` and use `default-sa` for its head node
- Set `enableInTreeAutoscaling` to true for the Ray cluster you just created
- The Kuberay operator will restart the head node pod, inject the autoscaler container, and use `foo` as its service account   

### Anything else

_No response_

### Are you willing to submit a PR?

- [x] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by davidxia**
## Related issue number

closes #242

## Checks

- [x] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by davidxia** [comment]
Do I need to add a test for this? If so, I'll need some pointers. 🙏

File: ray-operator/controllers/raycluster_controller.go [hunk 1]
```diff
@@ -578,7 +578,7 @@ func (r *RayClusterReconciler) reconcileAutoscalerServiceAccount(instance *rayio
 	}
 
 	serviceAccount := &corev1.ServiceAccount{}
-	namespacedName := types.NamespacedName{Namespace: instance.Namespace, Name: instance.Name}
+	namespacedName := types.NamespacedName{Namespace: instance.Namespace, Name: instance.Spec.HeadGroupSpec.Template.Spec.ServiceAccountName}
```
Comment (Jeffwan) [inline ray-operator/controllers/raycluster_controller.go:581]: 1. here, if the sa exists, it skip creating the new one.
2. you need to make sure head pod uses this service account name. -> I feel this part is missing?
@davidxia

File: ray-operator/controllers/common/rbac_test.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestGetHeadGroupServiceAccountName(t *testing.T) {
+	tests := map[string]struct {
+		input *rayiov1alpha1.RayCluster
+		want  string
+	}{
```
Comment (davidxia) [inline ray-operator/controllers/common/rbac_test.go:16]: Using table driven test with sub tests as described [here](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests).

File: ray-operator/controllers/common/rbac.go [hunk 3]
```diff
@@ -7,11 +7,21 @@ import (
...
 
+// GetHeadGroupServiceAccountName returns the head group service account if it exists.
+// Otherwise, it returns the name of the cluster itself.
+func GetHeadGroupServiceAccountName(cluster *v1alpha1.RayCluster) string {
+	headGroupServiceAccountName := cluster.Spec.HeadGroupSpec.Template.Spec.ServiceAccountName
+	if headGroupServiceAccountName != "" {
+		return headGroupServiceAccountName
+	}
+	return cluster.Name
+}
```
Comment (davidxia) [inline ray-operator/controllers/common/rbac.go:18]: utility function. Lmk if I should put elsewhere.

File: ray-operator/controllers/raycluster_controller.go (hunk 1 above)
Comment (davidxia) [reply ray-operator/controllers/raycluster_controller.go]: Thanks. I think I fixed 1 [here](https://github.com/ray-project/kuberay/pull/259/files#diff-37fdcfea947826e2a73f5ddcc0691ede6b513d411d6be53c3e529b950205e981R24) and 2 [here](https://github.com/ray-project/kuberay/pull/259/files#diff-8126a5e942d7e471be0f6a8e07cd26a9f4b9380b80c63ffc623f2ce33725a524R47).

File: ray-operator/controllers/common/pod.go [hunk 4]
```diff
@@ -9,6 +9,7 @@ import (
 	rayiov1alpha1 "github.com/ray-project/kuberay/ray-operator/api/raycluster/v1alpha1"
 	"github.com/ray-project/kuberay/ray-operator/controllers/utils"
 
+	"github.com/ray-project/kuberay/ray-operator/controllers/common"
```
Comment (davidxia) [inline ray-operator/controllers/common/pod.go:12]: My editor is complaining of a circular import here. Lmk if it's indeed a problem and how best to resolve. Probably by moving `GetHeadGroupServiceAccountName()` somewhere else?

**Comment by davidxia** [comment]
This is ready for review now

File: ray-operator/controllers/utils/util.go [hunk 5]
```diff
@@ -165,3 +166,64 @@ func FilterContainerByName(containers []corev1.Container, name string) (corev1.C
...
+// Otherwise, it returns the name of the cluster itself.
+func GetHeadGroupServiceAccountName(cluster *rayiov1alpha1.RayCluster) string {
+	headGroupServiceAccountName := cluster.Spec.HeadGroupSpec.Template.Spec.ServiceAccountName
+	if headGroupServiceAccountName != "" {
+		return headGroupServiceAccountName
+	}
+	return cluster.Name
+}
+
+func TestGetHeadGroupServiceAccountName(t *testing.T) {
```
Comment (Jeffwan) [inline ray-operator/controllers/utils/util.go:180]: Let's move test to https://github.com/ray-project/kuberay/blob/master/ray-operator/controllers/utils/util_test.go?

**Comment by Jeffwan** [comment]
The change looks good to me and please help move test to _test.go. Thanks! @davidxia

**Comment by davidxia** [comment]
> The change looks good to me and please help move test to _test.go. Thanks! @davidxia

@Jeffwan whoops, thanks. Done!

**Comment by davidxia** [comment]
@Jeffwan lmk next steps if any that are required for merging this. Thanks! 🙏

**Review by Jeffwan — APPROVED** [review]

**Comment by Jeffwan** [comment]
Thanks for the contribution!

**Pull request merged**